lib/src/working_copy.rs

```diff
@@ -680,19 +681,59 @@ impl TreeState {
                     }
 
                     if file_type.is_dir() {
-                        // If the whole directory is ignored, skip it unless we're already tracking
-                        // some file in it.
-                        if git_ignore.matches_all_files_in(&sub_path.to_internal_dir_string())
-                            && current_tree.path_value(&sub_path).is_none()
-                        {
-                            continue;
+                        if git_ignore.matches_all_files_in(&sub_path.to_internal_dir_string()) {
+                            // If the whole directory is ignored, visit only paths we're already
+                            // tracking.
+                            let tracked_paths = self
+                                .file_states
+                                .range((Bound::Excluded(&sub_path), Bound::Unbounded))
+                                .take_while(|(sub_path, _)| sub_path.contains(sub_path))
```

Comment:
`sub_path` shadows the outer `sub_path`

Reply:
I had renamed the outer one to `path` for a while then renamed it back to `sub_path`. I'm glad you noticed. I'll rename the outer one back to `path`.

Reply:
That's now done. I inserted a commit before the last one, in case you want to take another look before I merge this.